### ISSUE TITLE
No entry in device_lists left when user rejects invite

### ISSUE THREAD

**Issue by dbkr** [matrix-org/synapse#6163]
In https://github.com/matrix-org/synapse/blob/master/synapse/handlers/sync.py#L1392 we only include a member in `newly_left_users` if their previous membership was `join`, but in the other branch of the `if`, we include them in `newly_joined_or_invited_users` if they're, well, joined or invited, so if a user goes from `invite` to `leave` (or `ban`) then we'll never send them in the `left` section.

NB. clients shouldn't be relying on this to work out who to encrypt to (and indeed the js-sdk does not) so this doesn't cause events to be encrypted to anyone they shouldn't be. It might just cause clients to continue tracking the device lists of some users when they don't need to.

**Comment by erikjohnston** [comment]
Do we actually track devices for users we only share invites for? I didn't think we did? In which case I'm not sure if this is an actual bug

**Comment by DMRobertson** [comment]
From what I can see, this is handled to day here: https://github.com/matrix-org/synapse/blob/1eed795fc56d95df3968e37f3a4db92f24513e15/synapse/handlers/sync.py#L1399-L1425

Where newly_left_users is computed here: https://github.com/matrix-org/synapse/blob/1eed795fc56d95df3968e37f3a4db92f24513e15/synapse/handlers/sync.py#L1846-L1859

Specifically, here: https://github.com/matrix-org/synapse/blob/1eed795fc56d95df3968e37f3a4db92f24513e15/synapse/handlers/sync.py#L2547-L2578

**Comment by DMRobertson** [comment]
In the last quoted bit of source code you can see what David B is talking about.

- If some membership of A transitioned `* -> join`, `* -> invite`, or `* -> knock` during the sync period, then A is included in `newly_joined_or_invited_or_knocked_users`.
  - This also includes profile changes (`join -> join`), which seems unnecessary.
  - I don't know if `invite -> invite` or `knock -> knock` transitions are permitted (the [spec](https://spec.matrix.org/v1.5/client-server-api/#room-membership) is, as ever, vague) but they'd also be captured here.
- If there are no such transitions for A, but there is a transition `join -> leave` or `join -> ban`, A is included in `newly_left_users`.
  - Transitions `knock ->leave/ban` and `invite -> leave/ban` don't trigger this.
- There is no situation where a user belongs to both `newly_joined_or_invited_or_knocked` and `newly_left`.
- If A joins a room and leaves that room in the same sync period, they'll show up in `newly_joined_or_invited_or_knocked` but not `newly_left`, which is a surprise.

But it's worth noting that there's additional logic here which forms the /sync response: https://github.com/matrix-org/synapse/blob/1eed795fc56d95df3968e37f3a4db92f24513e15/synapse/handlers/sync.py#L1471-L1479

TL;DR it's surprising and confusing but I can't spot anything that's obviously wrong?
